Fix generating base keys from empty indexing paging state

psarna opened the pull request:
An empty partition/clustering key pair is a valid state of the 
query paging state. Unfortunately, recent attempts at debugging
a flaky test (#5856) resulted in introducing an assertion (7616290)
which breaks when trying to generate a key from such a pair.
In order to keep the assertion (since it still makes sense in its 
scope), but at the same time translate empty keys properly,
empty keys are now explicitly processed at the beginning of the 
function.
This behaviour was 100% reproducible in a secondary index dtest below.

Fixes #6134
Refs #5856
Tests: unit(dev),
     dtest(TestSecondaryIndexes.test_truncate_base)

nyh commented on cql3/statements/select_statement.cc:
```diff
@@ -514,25 +514,26 @@ indexed_table_select_statement::do_execute_base_query(
-                    command->slice.set_range(*_schema, base_pk, std::vector<query::clustering_range>{query::clustering_range::make_open_ended_both_sides()});
-                }
-            }
+            auto command = ::make_lw_shared<query::read_command>(*base_command);
```
Something looks wrong to me here - or maybe I just don't understand here.
In the old code, for the first iteration (concurrency==1), the code copied "cmd" and modified it based on the old paging state. But when concurrency > 1, the code also copied "cmd", but didn't modify it.
In the new code, all iterations copy the same base_command. What am I missing?
Also, do we have any test that can verify that this code actually works as intended?

psarna replied:
Oh snap, you're right... that's what I get for pushing patches too late in the evening. Only the first command should be trimmed, the rest should be a copy of the original command and this patch does not make sense. My original (correct) intention was that if we have an old paging state and our query spans multiple partitions, the first subquery may only need to fetch a small slice of the partition, depending of where we ended last time, but all the next queries should use the original, full range. I remember that there are dtests for that, I'll try to find out which ones.

psarna commented:
v2 changes:
 * removed the buggy "extraction out of the loop" commit
 * added regression unit test cases - which are verified to fail before the patch and succeed after the patch

nyh commented:
Committed to next.

eliransin commented:
@nyh @psarna , AFAICU there is no risk to backport it to 4.0 right? But should we also backport 
b3fdb742aed5ef1b8d849ef099a1e2044561831e before that? or should we avoid both of them altogether?

psarna commented:
Avoiding both is also ok. It will mean that we'll lose some debuggability *if* there's another problem with generating base keys, but it's not critical to backport.
